### ISSUE TITLE
prerequisites to get trace details from java application (spring bootstrap) to Zipkins

### ISSUE THREAD

**Issue by JasmithaMeka** [envoyproxy/envoy#2337]
*Title*: *prerequisites to get trace details from java application (spring bootstrap) to Zipkins*

*Description*:
Hi, I'm trying to understand prerequisites to get trace details from java application (spring bootstrap) to Zipkins. I've deployed two Microservices - one as API provider and other as API consumer in K8s cluster with Istio. Envoy is injected into pods. Zipkins is deployed as pod. I didn't get trace details in Zipkins though I generated traffic. I understand from IBM book info demo that few headers (x-request-id, x-b3-traceid, x-b3-spanid, x-b3-parentspanid, x-b3-sampled, x-b3-flags and x-ot-span-context) need to be propagated across . How are these headers are set in first place? 

Here is the piece of code for API produces microservice. When I deploy it along with side car and check logs, I'm getting null value for x-request-id and x-b3-traceid. What am I missing?

```
package microservice.one.greeting;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;

import javax.servlet.http.HttpServletRequest;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class GreetingController {

    private static final String template = "Hello from %s!";
    private final AtomicLong counter = new AtomicLong();
    
	@Autowired
	**private HttpServletRequest request;**

 
    @RequestMapping("/greeting")
    public Greeting greeting(@RequestHeader(value="myname") String header,@RequestParam(value="name", defaultValue="api-one") String name) {
    	System.out.println(request.getIntHeader("x-request-id"));
    	System.out.println(request.getHeader("user-agent"));
    	List<String> list=new ArrayList<String>();
    	list.add(request.getHeader("user-agent"));
    	**list.add(request.getHeader("x-request-id"));
    	list.add(request.getHeader("x-b3-traceid"));**
    	list.add(header);
    	
    	return new Greeting(counter.incrementAndGet(),
                            String.format(template, name),list);
        
    }
}
```

Snippet of API consumer

```
Greeting greeting = restTemplate.postForObject("http://192.168.33.12:30087/greeting",entity, Greeting.class);
				
				
				
System.out.println("----------------------------------------------");
System.out.println(greeting.getContent());
System.out.println(greeting.getId());
System.out.println(greeting.getHeader());		
System.out.println("----------------------------------------------");


```

This is the response I see on API consumer pod logs
```
----------------------------------------------
Hello from api-one!
159
[Java/1.8.0_151, null, null, jasmi]
----------------------------------------------
```

**Comment by mattklein123** [comment]
@adriancole @rshriram @fabolive any chance you could answer this question?

**Comment by rshriram** [comment]
Getting someone more familiar with Java to answer..

**Comment by adriancole** [comment]
TL;DR; This is a basic http controller that isn't even seeing a request ID. Maybe there's some basic configuration setup missing. Maybe refer the person to a chat support avenue if there's no checklist to dole out?

---

So the user is asking why their rest controller presumably invoked from envoy doesn't have trace headers present when viewing an incoming request.

So, I suppose it is a general troubleshooting question, for example is there any configuration that might have been missed, or are there logs one can view to verify that the endpoint hit indeed should have had trace IDs. Notably, there's not even a request ID present, which I assume any envoy originated request would have. Is there a way to suggest to the user how to verify the request details made it to the downstream service (ex via access logs or the like), just in case the endpoint was hit directly or otherwise out-of-band from envoy.

@kahootali mentioned OOB seems to work through a gitter conversation:

> @JasmithaMeka I have used Istio with Zipkin, it works fine actually all the network calls happen through Envoy SideCar so side car adds the headers, check if the pods are having a sidecar with them
> In case of Istio, you dont need Instrumentation libraries because sidecar does it on its own
https://istio.io/docs/tasks/telemetry/distributed-tracing.html#generating-traces-using-the-bookinfo-sample

**Comment by JasmithaMeka** [comment]
@adriancole Thanks for elaborating my query

**Comment by kahootali** [comment]
@JasmithaMeka I have tried in .NetCore, and printed the Ids, they are being printed.  Maybe you are missing something. Check this.
[Zipkin - Istio Demo](https://github.com/kahootali/Zipkin-IstioDemo)

**Comment by JasmithaMeka** [comment]
@kahootali thanks. It would be great if there is a demo in Java.

**Comment by adriancole** [comment]
probably we need to get an istio(?) example that uses the normal zipkin
frontend/backend.

https://github.com/openzipkin/brave-webmvc-example

I don't have cycles for this right now, though ideally the istio/envoy
setup really doesn't have any bearing with regards to the language used.

**Comment by JasmithaMeka** [comment]
Yes. I'm looking for a combination of Istio/Envoy + Zipkins + Java Application (without tracer library) right now.  I'm stressing on Java because for whatever application I am going to implement this is written in Java and it would helpful for me to understand dependencies and correlate easily.

I've also tried using Brave tracer in sample Java application. It works great, I could see traces in detail. I'm trying for Java Application where tracer is not included to see what all details do I get in traces(with Envoy) without change in code (Right now code does not include brave tracer). But somehow I don't get traces.

**Comment by adriancole** [comment]
Ok so you actually want a basic java app that isnt traced.. so you want the
same demo as was above (or similar) but instead of a dotnet service a java
one? It would seem strange there is no istio example already out there like
this. Maybe ask directly on an istio forum? I think you can ask more simply
for an istio java example.. maybe look for ray tsang from google's stuff.
Once that is working then try to enable tracing.

Good luck

**Comment by JasmithaMeka** [comment]
Yes exactly I'm looking for distributed tracing demo on simple Java application (without tracer) . 

Thanks

**Comment by mattklein123** [comment]
I'm going to close this out as I don't think this is an Envoy issue, per say. If you want to open a new issue to track adding an explicit Java example/demo please do.
